File: sdks/python/src/trustlayer/tracer.py
"""High-level tracer that ties events to an agent + session."""

from __future__ import annotations

import time
from collections.abc import Iterator
from contextlib import contextmanager
from typing import Any
from uuid import uuid4

from .client import TrustLayerClient
from .guardian import GuardianClient, Verdict
from .schema import (
    AgentTraceEvent,
    CynefinDomain,
    EventType,
    Metrics,
    PolicyCheckPayload,
    PolicyCheckResult,
    ToolCallPayload,
    ToolResultPayload,
)


class Tracer:
    """Bind an ``agent_id``/``session_id`` and emit typed trace events.

    Most callers will use :meth:`tool_call` as a context manager:

        tracer = Tracer(agent_id="researcher")
        with tracer.tool_call("web.search", {"q": "trustlayer"}) as out:
            out["value"] = run_search(...)
    """

    def __init__(
        self,
        agent_id: str,
        session_id: str | None = None,
        client: TrustLayerClient | None = None,
        cynefin_domain: CynefinDomain = CynefinDomain.DISORDER,
    ) -> None:
        self.agent_id = agent_id
        self.session_id = session_id or str(uuid4())
        self.client = client or TrustLayerClient()
        self.cynefin_domain = cynefin_domain

    def emit(
        self,
        event_type: EventType,
        payload: dict[str, Any] | None = None,
        metrics: Metrics | None = None,
        cynefin_domain: CynefinDomain | None = None,
    ) -> AgentTraceEvent:
        event = AgentTraceEvent(
            agent_id=self.agent_id,
            session_id=self.session_id,
            event_type=event_type,
            cynefin_domain=cynefin_domain or self.cynefin_domain,
            payload=payload or {},
            metrics=metrics or Metrics(),
        )
        self.client.emit(event)
        return event

    @contextmanager
    def tool_call(
        self,
        tool_name: str,
        tool_args: dict[str, Any] | None = None,
    ) -> Iterator[dict[str, Any]]:
        """Emit ``TOOL_CALL`` then ``TOOL_RESULT`` around a block.

        Set ``out["value"]`` inside the block to attach the tool's return
        value to the result event.
        """
        self.emit(
            EventType.TOOL_CALL,
            payload=ToolCallPayload(
                tool_name=tool_name, tool_args=tool_args or {}
            ).model_dump(),
        )
        start = time.perf_counter()
        out: dict[str, Any] = {}
        try:
            yield out
        except Exception as exc:
            self.emit(
                EventType.TOOL_RESULT,
                payload=ToolResultPayload(
                    tool_name=tool_name, error=repr(exc)
                ).model_dump(),
                metrics=Metrics(latency_ms=(time.perf_counter() - start) * 1000),
            )
            raise
        self.emit(
            EventType.TOOL_RESULT,
            payload=ToolResultPayload(
                tool_name=tool_name, result=out.get("value")
            ).model_dump(),
            metrics=Metrics(latency_ms=(time.perf_counter() - start) * 1000),
        )

    def policy_check(
        self,
        policy_name: str,
        action: str,
        result: PolicyCheckResult,
        reason: str | None = None,
    ) -> AgentTraceEvent:
        return self.emit(
            EventType.POLICY_CHECK,
            payload=PolicyCheckPayload(
                policy_name=policy_name,
                action=action,
                result=result,
                reason=reason,
            ).model_dump(),
        )

    def check(
        self,
        tool_name: str,
        tool_args: dict[str, Any] | None = None,
        *,
        guardian: GuardianClient,
        policy_name: str | None = None,
        cynefin_domain: CynefinDomain | None = None,
    ) -> Verdict:
        """Ask the guardian to adjudicate a candidate ``TOOL_CALL`` and record the verdict.

        Emits a ``TOOL_CALL`` event, forwards it to the guardian, and emits a
        follow-up ``POLICY_CHECK`` carrying the verdict. The caller inspects
        the returned :class:`Verdict` and decides whether to actually invoke
        the tool (typically: only on ``PASS``).

        The event sent to the guardian is the same one that was emitted,
        so the trace and the policy decision are correlated by ``trace_id``.
        """
        candidate = self.emit(
            EventType.TOOL_CALL,
            payload=ToolCallPayload(
                tool_name=tool_name, tool_args=tool_args or {}
            ).model_dump(),
            cynefin_domain=cynefin_domain,
        )
        verdict = guardian.check(candidate, policy_name=policy_name)
        self.policy_check(
            policy_name=verdict["policy"],
            action=f"invoke {tool_name}",
            result=PolicyCheckResult(verdict["decision"]),
            reason=verdict["reason"],
        )
        return verdict

File: sdks/go/trustlayer/tracer.go
package trustlayer

import (
	"context"
	"time"
)

// Tracer is a goroutine-safe convenience wrapper that knows which
// (agent_id, session_id) it belongs to and ships events through a
// shared TrustLayerClient.
type Tracer struct {
	client    *TrustLayerClient
	agentID   string
	sessionID string
}

// NewTracer pairs a client with an (agent_id, session_id).
func NewTracer(client *TrustLayerClient, agentID, sessionID string) *Tracer {
	return &Tracer{
		client:    client,
		agentID:   agentID,
		sessionID: sessionID,
	}
}

// Emit ships a pre-built event. Convenience pass-through to the client.
func (t *Tracer) Emit(ctx context.Context, event AgentTraceEvent) error {
	return t.client.Emit(ctx, event)
}

// ToolCall opens a tool-call span. Returns a closer that emits a
// TOOL_RESULT event when invoked (typically via defer). The result and
// error pointers are read at closer time, so callers can write to them
// in the body of the spanned code:
//
//	var result any
//	var err error
//	close := tracer.ToolCall(ctx, "calc", map[string]any{"x": 1}, &result, &err)
//	defer close()
//	result, err = doTheCall()
//
// If result/err are nil they're treated as "no result" / "no error".
func (t *Tracer) ToolCall(ctx context.Context, name string, args map[string]any, result *any, errOut *error) func() {
	startEvent := NewEvent(t.agentID, t.sessionID, EventToolCall,
		WithPayload(map[string]any{
			"tool_name": name,
			"tool_args": args,
		}),
	)
	start := time.Now()
	_ = t.client.Emit(ctx, startEvent)
	return func() {
		latencyMs := float64(time.Since(start).Microseconds()) / 1000.0
		payload := map[string]any{"tool_name": name}
		if result != nil {
			payload["result"] = *result
		}
		if errOut != nil && *errOut != nil {
			payload["error"] = (*errOut).Error()
		}
		resultEvent := NewEvent(t.agentID, t.sessionID, EventToolResult,
			WithPayload(payload),
			WithMetrics(Metrics{LatencyMs: &latencyMs}),
		)
		_ = t.client.Emit(ctx, resultEvent)
	}
}

// TracerCheck configures one call to Tracer.Check.
type TracerCheck struct {
	// Guardian is required.
	Guardian *GuardianClient
	// PolicyName overrides the guardian's configured default for this call.
	PolicyName string
	// CynefinDomain on the emitted TOOL_CALL + POLICY_CHECK events.
	// Defaults to CynefinDisorder if zero.
	CynefinDomain CynefinDomain
}

// Check is the ergonomic helper: emit a TOOL_CALL, ask the guardian,
// emit a POLICY_CHECK carrying the verdict so the trace and the
// decision share a trace_id, and return the Verdict so the caller can
// decide whether to invoke the tool. Matches the Python / TypeScript
// Tracer.check() shape (Phase 4.5).
func (t *Tracer) Check(ctx context.Context, toolName string, toolArgs map[string]any, opts *TracerCheck) (Verdict, error) {
	if opts == nil || opts.Guardian == nil {
		return Verdict{}, errMissingGuardian
	}
	domain := opts.CynefinDomain
	if domain == "" {
		domain = CynefinDisorder
	}
	call := NewEvent(t.agentID, t.sessionID, EventToolCall,
		WithCynefin(domain),
		WithPayload(map[string]any{
			"tool_name": toolName,
			"tool_args": toolArgs,
		}),
	)
	_ = t.client.Emit(ctx, call)

	verdict, _ := opts.Guardian.CheckWithPolicy(ctx, call, opts.PolicyName)

	policyName := opts.PolicyName
	if policyName == "" {
		policyName = verdict.Policy
	}
	reason := ""
	if verdict.Reason != nil {
		reason = *verdict.Reason
	}
	policyCheck := NewEvent(t.agentID, t.sessionID, EventPolicyCheck,
		WithCynefin(domain),
		WithPayload(map[string]any{
			"policy_name": policyName,
			"action":      toolName,
			"result":      string(verdict.Decision),
			"reason":      reason,
		}),
	)
	// Share trace_id with the originating tool call so the trace stream
	// can correlate the verdict with the action.
	policyCheck.TraceID = call.TraceID
	_ = t.client.Emit(ctx, policyCheck)
	return verdict, nil
}

// errMissingGuardian is returned when Tracer.Check is called without a
// Guardian wired up in TracerCheck.
var errMissingGuardian = &tracerError{msg: "trustlayer: Tracer.Check requires opts.Guardian"}

type tracerError struct{ msg string }

func (e *tracerError) Error() string { return e.msg }
